File: tree/path_sum/124_binary_tree_maximum_path_sum.cc
#include <climits>
#include <iostream>

#include "tree_node.h"

// 2024/01/14: 不看答案自己实现.

int res = INT_MIN;
int MaxPathSumUseRoot(TreeNode* root) {
  if (root == nullptr) {
    return 0;
  }
  int left = MaxPathSumUseRoot(root->left);
  int right = MaxPathSumUseRoot(root->right);
  res = std::max(res,
                 root->val + (left > 0 ? left : 0) + (right > 0 ? right : 0));
  if (left < 0 && right < 0) {
    return root->val;
  }
  return root->val + (left > right ? left : right);
}

int MaxPathSum(TreeNode* root) {
  MaxPathSumUseRoot(root);
  return res;
}

int main(int argc, char* argv[]) {
  TreeNode* node1 = new TreeNode(1);
  TreeNode* node2 = new TreeNode(2);
  TreeNode* node3 = new TreeNode(3);
  node1->left = node2;
  node1->right = node3;
  std::cout << MaxPathSum(node1) << std::endl;
  return 0;
}

File: tree/path_sum/124_binary_tree_maximum_path_sum.py
from typing import Optional
from tree.tree_node import TreeNode, create_tree_node

res = float("-inf")


# 2026/05/20: 不看答案自己实现.
# 包括 root 节点且只选择左子树或者右子树其中一个的最大值.
def max_sum_include_root(root: Optional[TreeNode]) -> int:
    global res
    if root is None:
        return 0

    left_max_sum = max_sum_include_root(root.left)
    right_max_sum = max_sum_include_root(root.right)
    res = max(res, root.val + max(left_max_sum, 0) + max(right_max_sum, 0))
    return root.val + max(left_max_sum, right_max_sum, 0)


def max_path_sum(root: Optional[TreeNode]) -> int:
    max_sum_include_root(root)
    return res


# root = [-10, 9, 20, None, None, 15, 7]
# root = create_tree_node(root)
# print(max_path_sum(root))

root2 = [5, 4, 8, 11, None, 13, 4, 7, 2, None, None, None, 1]
root2 = create_tree_node(root2)
print(max_path_sum(root2))
